### PR TITLE
implemented date_format scalar function

### PR DESCRIPTION
**Pull request by mfelsche**
in order to conveniently format date values.

it uses the MySQL date_format syntax.

### CODE REVIEW

**Comment by dobe** [comment]
hi @mfelsche nice to see prs from you :)

i think we need to make this compatible with https://dev.mysql.com/doc/refman/5.1/en/date-and-time-functions.html#function_date-format since we have many users migrating from mysql

File: sql/src/main/java/io/crate/operation/scalar/TimeZoneParser.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.joda.time.DateTimeZone;
+
+import java.util.Locale;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * parses timezone strings to {@linkplain DateTimeZone} instances
+ *
+ * TODO: make singleton
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/TimeZoneParser.java]: todo?

**Comment by mfelsche** [comment]
Made the function compatible with MySQL ``date_format``.
Please take another look. :)

File: sql/src/main/java/io/crate/metadata/Scalar.java [hunk 2]
```diff
@@ -54,6 +56,24 @@ protected static boolean anyNonLiterals(Collection<? extends Symbol> arguments)
         return false;
     }
 
+    protected static boolean hasNullInputs(Input<Object>... args) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Scalar.java]: hm, i think varargs are not required here, since you always use it with (args)

**Comment by dobe** [comment]
ack

**Comment by yuchen99** [comment]
Hi gays
which version can use `date_format` function.
Elastic support `date_histogram`
such as 
```json
  {
        "_source":false,
          "aggs": {
            "book_time": {
              "date_histogram" : {
                "field" : "book_time",
                "interval" : "30m"
            }
            }
          }
        }
```
can crate support that?

**Comment by mfelsche** [comment]
crate 0.51.0, which just hit our testing repos (e.g. [here](https://cdn.crate.io/downloads/releases/crate-0.51.0.tar.gz)), contains the ``date_format`` scalar function.

A date histogram can be created using the ``date_trunc`` scalar function. Your example, grouped by hour,would be roughly like this in SQL:

```sql
SELECT date_trunc('hour', book_time) as hour, COUNT(*) 
FROM books 
WHERE book_time IS NOT NULL AND book_time > '2015-07-01' AND book_time < '2015-07-02'
GROUP BY 1 
ORDER BY 1;
```
You can create a histogram from every interval, doing a *custom* ``date_trunc``: 

* express the interval as millis: ``30m -> 1800000ms``
* get the modulo from your date: ``(book_time%1800000)``
* subtract that modulo from your date: ``book_time - (book_time%1800000)``

```sql
SELECT book_time - (book_time%1800000) as "30m", COUNT(*)
FROM books
WHERE book_time IS NOT NULL AND book_time > '2015-07-01' AND book_time < '2015-07-02'
GROUP BY 1 
ORDER BY 1;
```

If you want to format the dates with a crate version prior to 0.51.0, you can use the [format](https://crate.io/docs/stable/sql/scalar.html#format-format-string-parameter-parameter-returns-string) with the [date/time conversions](http://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html#dt) but formatting will significantly add to the query execution time as you group by the formatted value, which is quite expensive.

**Comment by mfelsche** [comment]
next time please use an github issue or IRC or our google group for questions like that, like this it is easier to find for other people. See here: https://crate.io/community/support/

**Comment by yuchen99** [comment]
Awesome！
Thanks for help,next time I will use IRC

**Pull request merged**